src/main/java/com/fasterxml/uuid/impl/TimeBasedEpochGenerator.java

```diff
@@ -0,0 +1,121 @@
+ * method field from the Unix Epoch timestamp source - the number of 
+ * milliseconds seconds since midnight 1 Jan 1970 UTC, leap seconds excluded
+ * <p>
+ * As all JUG provided implementations, this generator is fully thread-safe.
+ * Additionally it can also be made externally synchronized with other instances
+ * (even ones running on other JVMs); to do this, use
+ * {@link com.fasterxml.uuid.ext.FileBasedTimestampSynchronizer} (or
+ * equivalent).
+ *
+ * @since 3.1
+ */
+public class TimeBasedEpochGenerator extends NoArgGenerator
+{
+    
+    public static int BYTE_OFFSET_TIME_HIGH = 0;
+    public static int BYTE_OFFSET_TIME_MID = 4;
+    public static int BYTE_OFFSET_TIME_LOW = 7;
+    
+    /*
+    /**********************************************************************
+    /* Configuration
+    /**********************************************************************
+     */
+
+    /**
+     * Object used for synchronizing access to timestamps, to guarantee
+     * that timestamps produced by this generator are unique and monotonically increasings.
+     * Some implementations offer even stronger guarantees, for example that
+     * same guarantee holds between instances running on different JVMs (or
+     * with native code).
+     */
+    protected final UUIDTimer _timer;
+
+
+    /**
+     * Random number generator that this generator uses.
+     */
+    protected final Random _random;
+    
+    /*
+    /**********************************************************************
+    /* Construction
+    /**********************************************************************
+     */
+    
+    /**
+     * @param rnd Random number generator to use for generating UUIDs; if null,
+     *   shared default generator is used. Note that it is strongly recommend to
+     *   use a <b>good</b> (pseudo) random number generator; for example, JDK's
+     *   {@link SecureRandom}.
+     */
+    
+    public TimeBasedEpochGenerator(UUIDTimer timer, Random rnd)
+    {
+        if (rnd == null) {
+            rnd = LazyRandom.sharedSecureRandom(); 
+        }
+        _random = rnd;
+        _timer = timer;
+    }
+    
+    /*
+    /**********************************************************************
+    /* Access to config
+    /**********************************************************************
+     */
+
+    @Override
+    public UUIDType getType() { return UUIDType.TIME_BASED_EPOCH; }
+    
+    /*
+    /**********************************************************************
+    /* UUID generation
+    /**********************************************************************
+     */
+    
+    /* As timer is not synchronized (nor _uuidBytes), need to sync; but most
+     * importantly, synchronize on timer which may also be shared between
+     * multiple instances
+     */
+    @Override
+    public UUID generate()
+    {
+        final long rawTimestamp = _timer.getTimestampV7();
```

Comment:
@cowtowncoder I am still a moron when it comes to bit twiddling. Have no idea what I'm doing wrong here ;)

I've fallen and I can't get up